ir/value.cpp

```diff
@@ -218,38 +218,15 @@ StateValue Input::mkInput(State &s, const Type &ty, unsigned child) const {
     s.addUndefVar(move(var));
   }
 
-  // Some attributes generate poison rather than raise UB
-  expr np_from_attr(true);
-  if (ty.isPtrType()) {
-    Pointer p(s.getMemory(), val);
-
-    if (hasAttribute(ParamAttrs::NonNull))
-      np_from_attr = p.isNonZero();
-
-    bool hasDeref = hasAttribute(ParamAttrs::Dereferenceable) ||
-                    hasAttribute(ParamAttrs::ByVal);
-    bool hasDerefOrNull = hasAttribute(ParamAttrs::DereferenceableOrNull);
-    if (hasDeref || hasDerefOrNull) {
-      if (hasDeref)
-        s.addUB(p.isDereferenceable(attrs.getDerefBytes(), attrs.align));
-      if (hasDerefOrNull)
-        s.addUB(p.isDereferenceable(attrs.derefOrNullBytes, attrs.align)() ||
-                !p.isNonZero());
-    } else if (hasAttribute(ParamAttrs::Align))
-      np_from_attr &= p.isAligned(attrs.align);
-  }
-
-  if (attrs.poisonImpliesUB()) {
-    s.addUB(move(np_from_attr));
-    np_from_attr = true;
-  }
+  auto [UB, non_poison] = attrs.encode(s, {expr(val), expr(true)}, ty);
```

Comment:
the non_poison part should take the `never_poison ? true : expr::mkBoolVar(np_name.c_str())` bit from below I guess

Reply:
never_poison considers `attrs.poisonImpliesUB()` into account, so in terms of correctness, it is fine.
I think this form is better because we can optimize away creation of `np_..` SMT variables. Otherwise attrs.encode will call `addUB(np_...)`.

Reply:
Actually this code becomes clearer if the non_poison part takes `never_poison ? true : expr::mkBoolVar(..)` as suggested.  If the performance cost of having the extra variable is almost negligible, I'll change this.